# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [1]:
import pandas as pd
df = pd.read_csv("C:/Users/rafag/Downloads/emendas.csv")
df.head(10)

C:\Users\rafag\AppData\Local\Temp\ipykernel_39352\3818065435.py:2: DtypeWarning: Columns (0,3,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("C:/Users/rafag/Downloads/emendas.csv")


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Aqui, a gente carregou o banco de dados e mostrou as 10 primeiras linhas

Explique em uma frase o que você observa nas primeiras 10 linhas.

## 2) Quantas linhas e colunas o dataset possui? 

In [ ]:
df.shape

(88596, 28)

O dataset possui 88596 linhas e 28 colunas

Explique o que esses números significam.

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [6]:
df.columns

Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos'],
      dtype='object')

Através do comando acima, listamos todas as colunas, na qual representam algumas características das emendas (Ano da ementa, nome, município, valor, etc)

Escreva sua breve descrição das colunas, como se estivesse criando um dicionário de variáveis para que leitor entenda.

## 4) Existem valores nulos? quantos?


In [7]:
nulos = df.isnull().sum()
print(nulos)

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             0
Código UF IBGE                        0
UF                                    0
Região                                0
Código Função                         0
Nome Função                           0
Código Subfunção                      0
Nome Subfunção                        0
Código Programa                       0
Nome Programa                         0
Código Ação                           0
Nome Ação                             0
Código Plano Orçamentário             0
Nome Plano Orçamentário               0
Valor Empenhado                       0
Valor Liquidado                       0
Valor Pago                            0


Neste dataset, não existem valores nulos.

## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88596 entries, 0 to 88595
Data columns (total 28 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Código da Emenda                    88596 non-null  object
 1   Ano da Emenda                       88596 non-null  int64 
 2   Tipo de Emenda                      88596 non-null  object
 3   Código do Autor da Emenda           88596 non-null  object
 4   Nome do Autor da Emenda             88596 non-null  object
 5   Número da emenda                    88596 non-null  object
 6   Localidade de aplicação do recurso  88596 non-null  object
 7   Código Município IBGE               88596 non-null  object
 8   Município                           88596 non-null  object
 9   Código UF IBGE                      88596 non-null  int64 
 10  UF                                  88596 non-null  object
 11  Região                              88596 non-null  ob

Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.

Algumas variáveis estão como objects, mesmo sendo um número. O ideal seria estar como inteiro (int) ou float.
As variáveis relacionadas ao financeiro estão como object, mas deveriam estar como float, e os valores relacionados à códigos da emenda, quantidade, ano,  número da emenda, e outros, também estão como objetc, mas deveriam estar como int.

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [14]:
df_ordenado = df.sort_values(by="Ano da Emenda", ascending=False)
df_ordenado.head()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20JP,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0000,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,"1800000,00","36000,00","0,00","0,00","0,00","0,00"
84655,202541860014,2025,Emenda Individual - Transferências com Finalid...,4186,EDUARDO GOMES,14,TOCANTINS (UF),Sem informação,Sem informação,1700000,...,15F9,APRIMORAMENTO DA INFRAESTRUTURA DA POLICIA FED...,0000,APRIMORAMENTO DA INFRAESTRUTURA DA POLICIA FED...,"1000000,00","1000000,00","1000000,00","0,00","0,00","0,00"
84646,202541860005,2025,Emenda Individual - Transferências com Finalid...,4186,EDUARDO GOMES,5,TOCANTINS (UF),Sem informação,Sem informação,1700000,...,8535,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,0000,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,"1000000,00","0,00","0,00","0,00","0,00","0,00"
84647,202541860006,2025,Emenda Individual - Transferências com Finalid...,4186,EDUARDO GOMES,6,Nacional,Sem informação,Sem informação,-1,...,6148,ASSISTENCIA MEDICA QUALIFICADA E GRATUITA A TO...,0000,ASSISTENCIA MEDICA QUALIFICADA E GRATUITA A TO...,"500000,00","500000,00","500000,00","0,00","0,00","0,00"
84648,202541860007,2025,Emenda Individual - Transferências Especiais,4186,EDUARDO GOMES,7,MÚLTIPLO,Sem informação,Múltiplo,1700000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"16469856,43","16305158,43","16305158,43","0,00","0,00","0,00"


Explique: qual é o ano mais recente presente no dataset?

O dataset está atualizado até o ano de 2025

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [9]:
df.describe()

,Ano da Emenda,Código UF IBGE
count,88596.000000,8.859600e+04
mean,2019.155109,2.943594e+06
std,3.409323,1.264884e+06
min,2014.000000,-1.400000e+01
25%,2016.000000,2.300000e+06
50%,2019.000000,3.100000e+06
75%,2022.000000,3.500000e+06
max,2025.000000,5.300000e+06


O comando describe mostra apenas duas colunas, isso ocorre porque os dados das outras colunas estão como objetos (string)

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [11]:
df.tail()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
88591,202544840005,2025,Emenda Individual - Transferências Especiais,4484,ZUCCO,5,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"12307991,08","12184912,08","12184912,08","0,00","0,00","0,00"
88592,202544840006,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,6,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"988649,90","25592,00","25592,00","0,00","0,00","0,00"
88593,202544840007,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,7,Nacional,Sem informação,Sem informação,-1,...,2000,ADMINISTRACAO DA UNIDADE,0000,ADMINISTRACAO DA UNIDADE - DESPESAS DIVERSAS,"999914,73","155246,57","149728,52","0,00","0,00","0,00"
88594,202544840008,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,8,Nacional,Sem informação,Sem informação,-1,...,2723,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0000,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...","497151,34","0,00","0,00","0,00","0,00","0,00"
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20JP,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0000,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,"1800000,00","36000,00","0,00","0,00","0,00","0,00"


A última linha está completa, ou seja, sem valores nulos.

Interprete a completude da última linha.

## 9) existem valores duplicados no dataframe? Caso positivo, quantos? 

In [12]:
duplicatas = df.duplicated().sum()
print(duplicatas)

0


Existem 0 valores duplicados